zeebe/exporters/operate-exporter/src/main/java/io/camunda/zeebe/operate/exporter/ExportBatchWriter.java

```diff
@@ -0,0 +1,131 @@
+import io.camunda.zeebe.protocol.record.Record;
+import io.camunda.zeebe.protocol.record.ValueType;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Accumulates Operate entities of different types and provide the method to flush them in batch.
+ */
+public class ExportBatchWriter {
+
+  // static
+  private final Map<ValueType, List<ExportHandler<?, ?>>> handlers = new HashMap<>();
+
+  // dynamic
+  private final Map<Tuple<String, Class<?>>, Tuple<OperateEntity, ExportHandler<?, ?>>>
+      cachedEntities = new HashMap<>();
+  private final Map<String, DecisionInstanceEntity> cachedDecisionInstanceEntities =
+      new HashMap<>();
+
+  public void addRecord(final Record<?> record) {
+    // TODO: need to filter to only handle events
+    final ValueType valueType = record.getValueType();
+
+    handlers
+        .getOrDefault(valueType, Collections.emptyList())
+        .forEach(
+            handler -> {
+              // TODO: ugly, generic problem
+              final ExportHandler handler2 = (ExportHandler) handler;
+
+              if (handler.handlesRecord((Record) record)) {
+
+                final List<String> entityIds = handler.generateIds((Record) record);
+                entityIds.forEach(
+                    id -> {
+                      final Tuple<String, Class<?>> cacheKey =
+                          new Tuple<>(id, handler.getEntityType());
+
+                      final OperateEntity cachedEntity;
+
+                      final boolean alreadyCached = cachedEntities.containsKey(cacheKey);
+                      if (alreadyCached) {
+                        cachedEntity = cachedEntities.get(cacheKey).getLeft();
+                      } else {
+                        cachedEntity = handler.createNewEntity(id);
+                      }
+
+                      handler2.updateEntity(record, cachedEntity);
+
+                      // always store the latest handler in the tuple, because that is the one
+                      // taking care of flushing
+                      cachedEntities.put(cacheKey, new Tuple<>(cachedEntity, handler));
+                    });
+              }
+            });
+  }
+
+  public void flush(final NewElasticsearchBatchRequest batchRequest) throws PersistenceException {
+    // some handlers modify the same entity (e.g. list view flow node instances are
+    // updated from process instance and incident records)
+    //
+    // the handler that modified the entity last will also flush it
+    if (cachedEntities.size() > 0) {
+      for (final Tuple<OperateEntity, ExportHandler<?, ?>> entityAndHandler :
+          cachedEntities.values()) {
+        final OperateEntity entity = entityAndHandler.getLeft();
+        final ExportHandler handler = entityAndHandler.getRight();
+        handler.flush(entity, batchRequest);
+      }
+      batchRequest.execute();
+      reset();
+    }
+  }
+
+  public void reset() {
```

Comment:
is there a situation where multiple threads might call this? Or one thread calls reset() while another thread is in another method utilizing the cache?